x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/jdbc/JdbcPreparedStatement.java

```diff
@@ -271,12 +318,101 @@ public void setNClob(int parameterIndex, Reader reader, long length) throws SQLE
 
     @Override
     public void setSQLXML(int parameterIndex, SQLXML xmlObject) throws SQLException {
-        throw new SQLFeatureNotSupportedException("SQLXML not supported");
+        setObject(parameterIndex, xmlObject);
     }
 
     @Override
     public void setObject(int parameterIndex, Object x, int targetSqlType, int scaleOrLength) throws SQLException {
-        throw new UnsupportedOperationException("Object not implemented yet");
+        checkOpen();
+        
+        JDBCType targetJDBCType;
+        try {
+            // this is also a way to check early for the validity of the desired sql type
+            targetJDBCType = JDBCType.valueOf(targetSqlType);
+        } catch (IllegalArgumentException e) {
+            throw new SQLDataException(e.getMessage());
+        }
+        
+        // set the null value on the type and exit
+        if (x == null) {
+            setParam(parameterIndex, null, targetSqlType);
+            return;
+        }
+        
+        checkKnownUnsupportedTypes(x);
+        if (x instanceof Boolean
+                || x instanceof Byte
+                || x instanceof Short
+                || x instanceof Integer
+                || x instanceof Long
+                || x instanceof Float
+                || x instanceof Double
+                || x instanceof String) 
+        {
+            setParam(parameterIndex, 
+                    TypeConverter.convert(x, TypeConverter.fromJavaToJDBC(x.getClass()), DataType.fromJdbcTypeToJava(targetJDBCType)), 
+                    targetSqlType);
+        } else if (x instanceof Timestamp
+                || x instanceof Calendar
+                || x instanceof Date
+                || x instanceof LocalDateTime
+                || x instanceof Time
+                || x instanceof java.util.Date) {
+            if (targetJDBCType == JDBCType.TIMESTAMP ) {
+                // converting to {@code java.util.Date} because this is the type supported by {@code XContentBuilder} for serialization
+                java.util.Date dateToSet;
+                if (x instanceof Timestamp) {
+                    dateToSet = new java.util.Date(((Timestamp) x).getTime());
```

Comment:
That's round the fractional milliseconds. Are we sure that is right? I think it is worth a comment if we are. If not, worth thinking about some more.

Reply:
@nik9000 I did it like this since ES deals with dates in milliseconds since epoch. The parameters will be serialized using a java.util.Date when being sent to ES in queries.

Reply:
Right. I think maybe this is worth playing with some more and documenting any strange things that come out of it. Though I think it'd be fine to do in a followup.